### ISSUE TITLE
Implement docker secrets

### ISSUE THREAD

**Issue by pablomendezroyo** [dappnode/DNP_DAPPMANAGER#1119]
**Context**
There are many dappnode packages that use passwords, tokens, and JWT. The current approach is to hardcode this token on both sides, server, and client. 

This has not critical from a security point of view but it's not ideal. If someone has access to the docker container through the token hardcoded means that the security of the dappnode has been already broken. However, this token hardcoded is limiting other functionalities such as exposing the HTTP_ENGINE through HTTPS with a token generated dynamically.

**Approach**
Use docker secrets to generate passwords/tokens dynamically on demand. These tokens/passwords must be generated dynamically based on the [`docker-compose` syntax](https://docs.docker.com/compose/compose-file/compose-file-v3/#secrets) during the package installation.

See [docker secrets](https://docs.docker.com/engine/swarm/secrets/)

**Comment by Nabsku** [comment]
For this to work theres a few things that need to happen in my opinion.

Docker secrets only work in docker swarm mode which means:
- We need to enable docker swarm with `docker swarm init`
- Packages that depend on secrets need to be deployed with `docker stack deploy` because docker-compose doesn't use swarm secrets

Then we'll need a check (pre-installation) if the desired secret exists (eg. `docker secret ls | grep engine_jwt`) and if not -> create it (eg. `echo "jwt_here" | docker secret create engine_jwt -`).

Anything I'm missing?

**Comment by pablomendezroyo** [comment]
Having to implement docker swarm to be able to use docker secrets (which will be widely used by many dappnode packages) would be a big change and implies a lot of changes in how things work. Also looks like docker swarm mode was implemented at some point by docker, so we will have to figgure out the exact version of docker when it was included, and how many potential users will not have docker swarm by default (it was docker swarm classic before). Not close to use docker swarm but we must know in deep the implications of implementing docker swarm.

I have seen that there is this `docker compose secrets` (https://docs.docker.com/compose/compose-file/compose-file-v3/#secrets)  does not require docker swarm to work. My idea would be:

- There would be two types of dappnode packages related to secrets. Creators and consumers
- The creators would define in the manifest the secret name and which packages services (by ENS) are allowed to use it.
- The consumers would define in the manifest the secret name that would like to consume.
- At the time of installing the creator package, secrets will be created and implemented through the `docker-compose` secrets
- At the time of installing the consumer package, secrets will be added or not if the package and service is allowed

**Comment by Nabsku** [comment]
> I have seen that there is this docker compose secrets (https://docs.docker.com/compose/compose-file/compose-file-v3/#secrets) does not require docker swarm to work. 

Yes but those are not "real" secrets, those are "fake" secrets. See: https://github.com/docker/compose/pull/4368

We could then only use file-backed secrets, meaning that they'd need to be defined when building the package or somehow injected into the docker-compose up sequence. Then again, this means that we would have secrets floating around in plain-text (on the nodes drive..) which would miss the point of doing this

Maybe looking into third-party secret management is worth it (hashicorp vault comes to mind). I'll see if I can get something up and running.

**Comment by alexpeterson91** [comment]
> Then again, this means that we would have secrets floating around in plain-text (on the nodes drive..) which would miss the point of doing this

Not so sure. If the secret is in plain text on the node but still generated on the node not just copied from GitHub (that would totally miss the point) it still would allow what we even want this for. Yes long term it's how security should work but as it stands you need to break all the layers of security already built into dappnode to be able to access the filesystem with root access like that. So we'd have bigger problems if that was the primary concern. 
I thought the main point right now at least was to use docker secrets to be able to securely pass the JWT or credentials over exposed https so you can run clients on different servers/pcs/DAppNodes

**Comment by Nabsku** [comment]
> Not so sure. If the secret is in plain text on the node but still generated on the node not just copied from GitHub (that would totally miss the point) it still would allow what we even want this for. Yes long term it's how security should work but as it stands you need to break all the layers of security already built into dappnode to be able to access the filesystem with root access like that. 

I guess we can do it like that. I'm just thinking about what the workflow looks like.

The secret creator side is pretty easy: 
- Take the package Besu as creator. Define the following in the manifest:
  ``` 
  "secrets": ["engine_auth"],
  "secretConsumers": ["lighthouse.dnp.dappnode.eth", "teku.dnp.dappnode.eth, [...] "],
  ```
- If package with `secrets` in manifest is getting installed, automatically create secrets on local drive
- If secret with name defined in manifest exists already, do not create a new one  

  > ⚠️ Can't wait for consumer to get installed, since Besu - or any other EL client for that matter - itself needs the secret to work. Preferably set file permissions to `chmod 600` since only docker has to read the secrets (or am I missing something?)

- Write available secrets and eligible consumers to some sort of database/persistent storage (so we don't have to go through all manifests during consumer install)

But I'm wondering how we should implement the consumer side.. this is my current draft:

- Take Lighthouse as consumer
- Check if Lighthouse demands access to any secrets defined in the manifest:

  ```
   "requiredSecrets": ["engine_auth"],
  ```
- If consumer is allowed to use secret, place secret file in package storage location 

  `/var/lib/docker/volumes/dncore_dappmanagerdnpdappnodeeth_data/_data/lighthouse.dnp.dappnode.eth`

Anything I'm missing here @pablomendezroyo?

**Comment by pablomendezroyo** [comment]
> If package with secrets in manifest is getting installed, automatically create secrets on local drive
Here we have several options: 
- Create the secret dynamically in the dappmanager (without storing it in local drive), add it to the creator using [docker-compose secrets](https://docs.docker.com/compose/compose-file/compose-file-v3/#secrets) (no external ones are allowed to be used without docker swarm). Whenever a consumer is installed, the dappmanager will use `docker cp <container_creator/secret>` to retrieve the secret from the creator and add it to the consumer, also using docker-compose secrets, so it gets persisted.
- Having a new core package for handling secrets (as @Nabsku proposed) using a docker image such as [vault](https://hub.docker.com/_/vault). The main issue here is that the dappmanager is the only one who truly knows which can or cannot be the consumers by ENS filtering.

> If secret with name defined in manifest exists already, do not create a new one
This is one of many more edge cases, I would throw an error if the secret already exists. Also, if the creator gets removed the secrets should be removed as well

> Write available secrets and eligible consumers to some sort of database/persistent storage (so we don't have to go through all manifests during consumer install)
True, there must be some kind of database. But I would just store the creator and the secret name: `{ creator: geth.dnp.dappnode.eth, consumers: [lighthouse.dnp.dappnode.eth, prysm.dnp.dappnode.eth]}`. When installing a creator, there will be a search for the secret name and a comparison between the consumer ENS and the database ENS. If false error, if true then get the secret from the consumer with `docker cp`


This implementation would allow us to not persist the secret nowhere but only on the consumers/creators

**Comment by dsimog01** [comment]
We could use some kind of information only the host knows to generate a JWT in a way in which both the consensus client and the execution client reach to the same value. For example, we could do something like giving an index to each client (prysm=0, lighthouse=1...) and then `JWT=hash(hash(recovery_token)+client_index)`. If needed, we could get a substring of the result
